src/phi2.cpp

```diff
@@ -1,68 +1,133 @@
+  *       = asinh(tan(chi))
+  * chi is the conformal latitude
+  *
+  * The representation of latitudes via their tangents, tan(phi) and tan(chi),
+  * maintains full *relative* accuracy close to latitude = 0 and +/- pi/2.
+  * This is sometimes important, e.g., to compute the scale of the transverse
+  * Mercator projection which involves cos(phi)/cos(chi) tan(phi)
+  *
+  * From Karney (2011), Eq. 7,
+  *
+  *   tau' = sinh(psi) = sinh(asinh(tan(phi)) - e * atanh(e * sin(phi)))
+  *        = tan(phi) * cosh(e * atanh(e * sin(phi))) -
+  *          sec(phi) * sinh(e * atanh(e * sin(phi)))
+  *        = tau * sqrt(1 + sigma^2) - sqrt(1 + tau^2) * sigma
+  * where
+  *   sigma = sinh(e * atanh( e * tau / sqrt(1 + tau^2) ))
+  *
+  * For e small,
+  *
+  *    tau' = (1 - e^2) * tau
+  *
+  * The relation tau'(tau) can therefore by reliably inverted by Newton's
+  * method with
+  *
+  *    tau = tau' / (1 - e^2)
+  *
+  * as an initial guess.  Newton's method requires dtau'/dtau.  Noting that
+  *
+  *   dsigma/dtau = e^2 * sqrt(1 + sigma^2) /
+  *                 (sqrt(1 + tau^2) * (1 + (1 - e^2) * tau^2))
+  *   d(sqrt(1 + tau^2))/dtau = tau / sqrt(1 + tau^2)
+  *
+  * we have
+  *
+  *   dtau'/dtau = (1 - e^2) * sqrt(1 + tau'^2) * sqrt(1 + tau^2) /
+  *                (1 + (1 - e^2) * tau^2)
+  *
+  * This works fine unless tau^2 and tau'^2 overflows.  This may be partially
+  * cured by writing, e.g., sqrt(1 + tau^2) as hypot(1, tau).  However, nan
+  * will still be generated with tau' = inf, since (inf - inf) will appear in
+  * the Newton iteration.
+  *
+  * If we note that for sufficiently large |tau|, i.e., |tau| >= 2/sqrt(eps),
+  * sqrt(1 + tau^2) = |tau| and
+  *
+  *   tau' = exp(- e * atanh(e)) * tau
+  *
+  * So
+  *
+  *   tau = exp(e * atanh(e)) * tau'
+  *
+  * can be returned unless |tau| >= 2/sqrt(eps); this then avoids overflow
+  * problems for large tau' and returns the correct result for tau' = +/-inf
+  * and nan.
+  *
+  * Newton's method usually take 2 iterations to converge to double precision
+  * accuracy (for WGS84 flattening).  However only 1 iteration is needed for
+  * |chi| < 3.35 deg.  In addition, only 1 iteration is needed for |chi| >
+  * 89.18 deg (tau' > 70), if tau = exp(e * atanh(e)) * tau' is used as the
+  * starting guess.
+  ****************************************************************************/
+
+  constexpr int numit = 5;
+  // min iterations = 1, max iterations = 2; mean = 1.954
+  constexpr double tol = sqrt(std::numeric_limits<double>::epsilon()) / 10;
+  constexpr double tmax = 2 / sqrt(std::numeric_limits<double>::epsilon());
+  double
+    e2m = 1 - e * e,
+    tau = fabs(taup) > 70 ? taup * exp(e * atanh(e)) : taup / e2m,
+    stol = tol * std::max(1.0, fabs(taup));
+  if (!(fabs(tau) < tmax)) return tau; // handles +/-inf and nan and e = 1
+  // if (e2m < 0) return std::numeric_limits<double>::quiet_NaN();
+  int i = numit;
+  for (; i; --i) {
+    double tau1 = sqrt(1 + tau * tau),
```

Comment:
All these are const, yes?  I've seen too many sneaky things over the years.

Reply:
We went around the block a few times on this particular question a while back.  I know that I can make these variable `const`.  But that's not how *I* think of them which is "a bunch of variable I need in this do loop". `const` elevates them in importance so now I have to ask: "gee, is there something I need to watch out for?"  So in my mind, I reserve the `const` declaration for long-lived immutable quantities; so maybe I would agree with you if the boday of the for loop was 50 lines or more.  This is all a matter of taste, of course.  The compiler will produce the same code in either case.

Reply:
It's fine to not make them const, but I take issue with `const declaration for long-lived immutable quantities` and `elevates them in importance`.  e.g.  This says I'm iterating over something and I'm not mutating it.  It's definitely not important.

```
for (const auto& item : container) {
  // short bit of code.
}
```